app/scripts/main.js

```diff
@@ -28,53 +30,51 @@
 
     cameraManager.onframe = function() {
       // There is a frame in the camera, what should we do with it?
- 
-      var imageData = cameraManager.getImageData();
-      var detectedQRCode = qrCodeManager.detectQRCode(imageData, function(url) {
-        if(url !== undefined) {
-          qrCodeManager.showDialog(url);
-        }
-      });
+      if (qrCodeManager.numActiveThreads < qrCodeManager.maxThreads) {
+        qrCodeManager.detectQRCode(cameraManager.getImageData());
+        qrCodeManager.numActiveThreads++;
+      }
+    };
+
+    // Worker listener
+    imageDecoderWorker.onmessage = function (e) {
+      var url = e.data;
+      // freeing 1 thread slot
+      qrCodeManager.numActiveThreads--;
+
+      if (url) {
+        // found a QRCode
+        qrCodeManager.showDialog(url);
+        qrCodeManager.currentUrl = url;
+      }
+    };
     
+    imageDecoderWorker.onerror = function(error) {
+      function WorkerException(message) {
+        this.name = "WorkerException";
+        this.message = message;
+      };
+      throw new WorkerException('Decoder error');
     };
   };
 
   var QRCodeManager = function(element) {
+    this.numActiveThreads = 0;
+    this.maxThreads = 4;
```

Comment:
Just an arbitrary number - good question.
A better approach could be:

```js
  this.maxThreads = navigator.hardwareConcurrency || 4;
```

https://developer.mozilla.org/en-US/docs/Web/API/NavigatorConcurrentHardware/hardwareConcurrency

I'll update the PR tonight